### PR TITLE
Add PARAM_TYPE_SIZE

### PR DESCRIPTION
**Pull request by tristan957**
Adds support for having size_t parameters.

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

**Comment by tristan957** [comment]
From my research seems like size_t has to be guaranteed to be at least 16 bits in width, so I went ahead and just accounted for various scenarios in the test.

File: lib/config/param.c [hunk 1]
```diff
@@ -443,6 +445,40 @@ param_default_converter(const struct param_spec *ps, const cJSON *node, void *va
...
+            if (to_conv < 0 || *end != '\0') {
+                log_err(
+                    "Value of %s must be greater than or equal to 0 and less than or equal to %lu",
+                    ps->ps_name,
+                    SIZE_MAX);
+                return false;
+            }
+            *(size_t *)value = tmp;
+            break;
+        }
```
Comment (beckerg) [inline lib/config/param.c:481]: I still don't know why we need to do this..

FWIW, in order to use the strtoul() type functions correctly you need to set errno=0 before calling, then check errno after the call, then check to see if return code is ULONG_MAX and if see then check errno...
I don't think I'd trust that end is valid in that case...

File: lib/config/param.c [hunk 2]
```diff
@@ -921,6 +973,15 @@ param_default_validator(const struct param_spec *ps, const void *value)
...
                 break;                                                                            \
+            case PARAM_TYPE_SIZE:                                                                 \
+                if (tmp < 0 || tmp > (double)SIZE_MAX) {                                          \
+                    log_err(                                                                      \
+                        "Number of bytes of %s is not within the bounds of a size_t",             \
+                        ps->ps_name);                                                             \
+                    return false;                                                                 \
+                }                                                                                 \
+                *(size_t *)value = (size_t)tmp;                                                   \
+                break;                                                                            \
```
Comment (beckerg) [inline lib/config/param.c:984]: I'm suprised the compiler doesn't complain about lost of precision due to cast to double...

Comment (tristan957) [reply lib/config/param.c:984]: Perhaps it would if we had `-Wconversion` enabled

**Review by beckerg — APPROVED** [review]

File: lib/config/param.c (hunk 2 above)
Comment (beckerg) [reply lib/config/param.c:984]: Casting to (long double) should preserve all 64 bits, I think...

File: lib/config/param.c (hunk 1 above)
Comment (tristan957) [reply lib/config/param.c:481]: It can all go away when cJSON is removed.

Hopefully fixed the error checking in next push.

Comment (beckerg) [reply lib/config/param.c:491]: Cool, btw, shouldn't you use %lf in your snprintf?

Comment (tristan957) [reply lib/config/param.c:492]: Whoops, fixed.

**Comment by alexttx** [comment]
> From my research seems like size_t has to be guaranteed to be at least 16 bits in width, so I went ahead and just accounted for various scenarios in the test.

What is the benefit of using `size_t` for our param types instead of fixed width types?

**Comment by tristan957** [comment]
There are a few places in the code where we cast various size-based parameters to size_t. One particular case where this came up was a discussion in Greg's PR.

**Comment by tristan957** [comment]
You definitely bring up a good point. This PR can go into hibernate and resurrected at a different time if need be.

File: lib/config/lib/params.c [hunk 3]
```diff
@@ -415,6 +417,49 @@ param_default_converter(const struct param_spec *ps, const cJSON *node, void *va
...
+                return false;
+            }
+
+            /* Converting a double to a 64-bit size_t is extremely difficult.
+             * Confirm that the double is a whole number, convert it to a
+             * string, and parse the string to a size_t.
+             */
+            snprintf(buf, sizeof(buf), "%.0lf", to_conv);
+            errno = 0;
+#if SIZE_MAX == UINT64_MAX
```
Comment (alexttx) [inline lib/config/lib/params.c:445]: wouldn't `sizeof(size_t) == sizeof(unsigned long long)` be better here?  or a static assert that sizeof `size_t` is <= size of `unsigned long long` and always use `unsigned long long`?

**Review by alexttx — APPROVED** [review]

File: lib/config/lib/params.c (hunk 3 above)
Comment (tristan957) [reply lib/config/lib/params.c:445]: I can't seem to use `sizeof` in the `#if`. Assuming that is what you mean of course.

Comment (tristan957) [reply lib/config/lib/params.c:445]: Remember that this code is also temporary till we get a better json library. All this string conversion will go the way of the dodo

**Pull request merged**